File: Python/Point2D.py
#!/usr/bin/python3

  # Point2D
  # Connexion

  # Created by Jalel Benerrami on 29/02/2016.
  # Copyright © 2016 Jalel Benerrami. All rights reserved.


import numpy as np
import matplotlib.pyplot as plt
import math

#
# A 2D point given by (x,y) coordinates

class Point2D :
    #
    # Create a new 2D point with given coordinates
    def __init__(self, x0, y0):
        self._x = x0
        self._y = y0
    #
    # Get coordinates
    def x(self):
        return self._x
    def y(self):
        return self._y
    #
    # Display a 2D point
    def __repr__(self):
        return "({},{})".format(self._x, self._y)
    def print(self):
        print(self)
    #
    # Calculate the Euclidian distance between two points
    def distance(self, pt):
        return math.sqrt((self.x()-pt.x())**2 + (self.y()-pt.y())**2)
    #
    # Draw a line between two points
    def connect(self, pt):
        abs = np.array([self.x(), pt.x()])
        ord = np.array([self.y(), pt.y()])
        plt.plot(abs, ord, "b-")
        # plt.plot(abs, ord, "bo-") to mark the points with symbol o
    #
    # Determine the nearest neighbourg of a point among the ones in a vector
    def nearestNeighbour(self, listPoints):
        nN = listPoints[0]
        for pt in listPoints:
            if(self.distance(pt) < self.distance(nN)):
                nN = pt
        return nN
    #
    # # Add a point to an existing list of points
    # def addPoint2D(self, listPoints):
    #     listPoints.append(self)
    

File: C++/Point2D.cpp
//
//  Point.cpp
//  Connexion
//
//  Created by Jalel Benerrami on 26/02/2016.
//  Copyright © 2016 Jalel Benerrami. All rights reserved.
//

#include "Point2D.h"
using std::ostream;
#include <math.h>    /* sqrt */




//
//Define a order between two Point2D
//
bool Point2D::operator<(const Point2D &pt) const {
    return _x < pt.x() || (_x == pt.x() && _y < pt.y());
}

//
// Calculate the Euclidian distance between two points
//
float Point2D::distance(const Point2D &pt) const {
    return sqrt(powf(_x-pt.x(),2) + powf(_y-pt.y(),2));
}

//
// Draw a line between two points
//
void Point2D::connect(const Point2D &pt, const Color &c) const {
    drawLine(_x,_y,pt.x(),pt.y(),c);
}

//
// Add a point to an existing list of points
//
void Point2D::addPoint2D(vector<Point2D> &v) const {
    v.push_back(*this);
}

//
// Determine the nearest neighbourg of a point among the ones in a vector
//
int Point2D::nearestNeighbour(const vector <Point2D> &v) const {
    int nN=0;
    for(vector<Point2D>::const_iterator it = v.begin(); it != v.end(); ++it){
        if(this->distance(v[nN]) > this->distance(*it)){
            nN = it - v.begin();
        }
    }
    return nN;
}

//
// Determine the nearest neighbourg of a point among the ones in a set
//
Point2D Point2D::nearestNeighbour(const set <Point2D> &s) const{
    Point2D nN = *(s.begin());
    for(set<Point2D>::const_iterator it = s.begin(); it != s.end(); ++it){
        if(this->distance(nN) > this->distance(*it)){
            nN = *it;
        }
    }
    return nN;
}

//
// Multiply or divide a point coordonnates by the same floating value
//
Point2D Point2D::homothety(const float &k, const float &l)const{
    return Point2D(_x/k,_y/l);
}

//
// Set the color
//
void Point2D::colorIn(const Color &c){
    _c = c;
}


//
// Display 2D point
//
ostream &operator<<(ostream &os, Point2D const &pt) {
    return os << "(" << pt.x() << "," << pt.y() << ")";
}
ostream &operator<<(ostream &os, Point2D* const &p) {
    return os << *p;
}